### DISCUSSION TITLE
how to get older version of setuptools

### DISCUSSION THREAD

**Discussion by lbk861** [pypa/setuptools#4354] [Q&A] [answered]
environment: Python 3.11.9 venv on Alpine Linux

I'm installing a package. The latest  setuptools already removed `_get_unpatched` which is needed by that package so I tried to degrade the setuptools by pip install  setuptools=18.3.2.  Still failed to pip install that.
Then I dircetly modified the setup.py    `setup_requires=['setuptools==18.3.2']` and run pip install. 
I get an error: `conflict with the backend dependencies: setuptools==69.5.1 is incompatible with setuptools==18.3.2.`

It seems that pip install doesn't use installed setuptools ?

**Comment by abravalheri** [comment]
Have you tried pinning the version in `pyproject.toml`'s `[build-system] requires = ...`?

(The `pyproject.toml` of the package you are trying to build/install).

**Reply by lbk861** [reply]
I don't find the pyproject.toml in it.

**Reply by abravalheri** [reply, marked as answer]
OK, in that case I suppose you can try `pip install --no-build-isolation --no-use-pep517 ...`. You will also need to install all build dependencies manually, that includes `wheel`.

But if the package is that old and relies on a very old version of `setuptools`, there is a chance you need an old version of `pip` to make it work too.

**Reply by lbk861** [reply]
> OK, in that case I suppose you can try pip install --no-build-isolation --no-pep517 ....

I try the command but it outputs `no such option: --no-pep517`. I search online and it may be `--no-use-pep517` ?
Also I try the `--no-use-pep517`, the result :
```
--no-use-pep517 error: It is not possible to use --no-use-pep517
without setuptools and wheel installed.
```

> But if the package is that old and relies on a very old version of setuptools, there is a chance you need an old version of pip to make it work too.

 `_get_unpatched` that is needed by the package is available when the version of setuptools is less than v68.0.0.

**Reply by abravalheri** [reply]
> --no-use-pep517 error: It is not possible to use --no-use-pep517
without setuptools and wheel installed.

Exactly, that is why I said previously `You will also need to install all build dependencies manually, that includes wheel`. You need to manually install the version of setuptools that you want + all the build dependencies that you need (and that includes `wheel`), before trying to install the package.

**Reply by lbk861** [reply]
Oh, yeah! Thanks! I think wheel is already installed before and actually only setuptools is installed😀.

And I'm confused that when pip install one package,  with build-isolation does pip always not  use installed dependents ?

**Reply by abravalheri** [reply]
In the future the idea is that `pip` will use build isolation by default. But for now, sometimes it does, sometimes is does not, it all depend on some internal `pip` heuristics.

You can read about build isolation in https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/#build-isolation, but it is basically equivalent of creating a brand new virtual environment just for running the build, and "re-installing" all the dependencies from scratch in this isolated build environment.

**Comment by abravalheri** [comment]
`setuptools==18.3.2` 😨
